activation/atxdb.go

```diff
@@ -316,111 +307,88 @@ func (db *ActivationDb) ActiveSetSize(epochId types.EpochId) (uint32, error) {
 	return common.BytesToUint32(val), nil
 }
 
+type atxIdAndLayer struct {
+	AtxId   types.AtxId
+	LayerId types.LayerID
+}
+
 // addAtxToEpoch adds atx to epoch epochId
 // this function is not thread safe and needs to be called under a global lock
-func (db *ActivationDb) addAtxToEpoch(epochId types.EpochId, atx types.AtxId) error {
-	ids, err := db.atxs.Get(epochId.ToBytes())
-	var atxs []types.AtxId
+func (db *ActivationDb) updatePosAtxIfNeeded(atx *types.ActivationTx) error {
+	currentIdAndLayer, err := db.getCurrentAtxIdAndLayer()
+	if err != nil && err != database.ErrNotFound {
+		return fmt.Errorf("failed to get current ATX ID and layer: %v", err)
+	}
+	if err == nil && currentIdAndLayer.LayerId >= atx.PubLayerIdx {
+		return nil
+	}
+
+	newIdAndLayer := atxIdAndLayer{
+		AtxId:   atx.Id(),
+		LayerId: atx.PubLayerIdx,
+	}
+	idAndLayerBytes, err := types.InterfaceToBytes(&newIdAndLayer)
 	if err != nil {
-		//epoch doesnt exist, need to insert new layer
-		ids = []byte{}
-		atxs = make([]types.AtxId, 0, 0)
-	} else {
-		atxs, err = decodeAtxIds(ids)
-		if err != nil {
-			return errors.New("could not get all atxs from database ")
-		}
+		return fmt.Errorf("failed to marshal posAtx ID and layer: %v", err)
+	}
+
+	err = db.atxs.Put([]byte(posAtxKey), idAndLayerBytes)
+	if err != nil {
+		return fmt.Errorf("failed to store posAtx ID and layer: %v", err)
+	}
+	return nil
+}
+
+func (db ActivationDb) getCurrentAtxIdAndLayer() (atxIdAndLayer, error) {
```

Comment:
What is this type/ function? why is that needed?

Reply:
I think that this optimisation is no needed, just read the atx, it's only a few more bytes

Reply:
The name of this function is unclear, I'll improve it: the "current ATX" is the ATX with the highest layer ID that we've ever seen. This ATX will be used as a positioning ATX.

Regarding the type, perhaps we can use the ATX header here instead of this new type. Didn't think introducing a type was a big deal.

Reply:
Actually, we can't use the ATX header, since it doesn't contain the data required to calculate the ID, nor does it include the ID.

We could use the full ATX (incl. NIPST and view) but that's more than a few more bytes, and it also requires calculating the ID, which is expensive.

What's wrong with using a new type for this?

Reply:
I've made naming improvements. I called the ATX with the highest layer, that we'll potentially use as a positioning ATX "top ATX".

I've also simplified and clarified some variable and function names as a result.

The code is clearer and easier to read now, IMO.